lib/rubocop/target_finder.rb

```
@@ -142,6 +143,10 @@ def without_excluded(files)
       end
     end
 
+    def reject_existing_non_files!(files)
+      files.reject! { |file| FileTest.exist?(file) && !FileTest.file?(file) }
```

This only happens when using globs, right? Doing such a check for every file in a project would be pretty expensive otherwise, especially when it barely comes up. You should also be able to get away with just doing `file?` here, it checks for existance already.

If it does happen for every file, then the better solution would be to just rescue the exception at the place it happens and assume it happens seldomly

---

Yep, it makes sense. I didn't consider performance impact of these changes, will try another approach

---

On the other hand, it seems like we already [do](https://github.com/rubocop/rubocop/blob/a8c001472d5d670798633d6c052d285e5f33d68a/lib/rubocop/target_finder.rb#L68) `FileTest.file?` if positional argument is not a glob

---

I guess it doesn't really matter then, my bad. Regardless, just `!FileTest.file?(file)` should be enough.

---

> Regardless, just !FileTest.file?(file) should be enough.

Not really, unfortunately. If you pass files paths (not directories), `TargetFinder` currently returns them even if these files do not exist. I'm not sure if it's intentional or not, so I added existence check to keep this Behaviour.

These tests fail after existence check removal:

```
rspec './spec/rubocop/cli/options_spec.rb[1:23:1]' # RuboCop::CLI options --stdin causes source code to be read from stdin
rspec './spec/rubocop/cli/options_spec.rb[1:23:8]' # RuboCop::CLI options --stdin detects CR at end of line
rspec './spec/rubocop/cli/options_spec.rb[1:23:4]' # RuboCop::CLI options --stdin prints corrected code to stdout if --autocorrect-all is used
rspec './spec/rubocop/cli/options_spec.rb[1:23:5]' # RuboCop::CLI options --stdin prints offense reports to stderr and corrected code to stdout if --autocorrect-all and --stderr are used
rspec './spec/rubocop/cli/suggest_extensions_spec.rb[1:1:15:1]' # RuboCop::CLI SuggestExtensions extension suggestions when given an invalid path does not show the suggestion
rspec './spec/rubocop/lsp/server_spec.rb[1:21:1]' # RuboCop::LSP::Server execute command safe formatting with `Lint/UnusedBlockArgument` cop (`AutoCorrect: contextual`) handles requests
rspec './spec/rubocop/lsp/server_spec.rb[1:6:1]' # RuboCop::LSP::Server format with `safeAutocorrect: false` handles requests
rspec './spec/rubocop/lsp/server_spec.rb[1:20:1]' # RuboCop::LSP::Server execute command safe formatting handles requests
rspec './spec/rubocop/lsp/server_spec.rb[1:13:1]' # RuboCop::LSP::Server format with `lintMode: true` and `layoutMode: true` handles requests
rspec './spec/rubocop/lsp/server_spec.rb[1:11:1]' # RuboCop::LSP::Server format with `layoutMode: true` handles requests
rspec './spec/rubocop/lsp/server_spec.rb[1:5:1]' # RuboCop::LSP::Server format with `safeAutocorrect: true` handles requests
rspec './spec/rubocop/lsp/server_spec.rb[1:9:1]' # RuboCop::LSP::Server format with `lintMode: false` handles requests
rspec './spec/rubocop/lsp/server_spec.rb[1:3:1]' # RuboCop::LSP::Server format by default (safe autocorrect) handles requests
rspec './spec/rubocop/lsp/server_spec.rb[1:2:1]' # RuboCop::LSP::Server did open handles requests
rspec './spec/rubocop/lsp/server_spec.rb[1:7:1]' # RuboCop::LSP::Server format without `lintMode` option handles requests
rspec './spec/rubocop/lsp/server_spec.rb[1:22:1]' # RuboCop::LSP::Server execute command unsafe formatting handles requests
rspec './spec/rubocop/lsp/server_spec.rb[1:8:1]' # RuboCop::LSP::Server format with `lintMode: true` handles requests
rspec './spec/rubocop/lsp/server_spec.rb[1:12:1]' # RuboCop::LSP::Server format with `layoutMode: false` handles requests
rspec './spec/rubocop/lsp/server_spec.rb[1:10:1]' # RuboCop::LSP::Server format without `layoutMode` option handles requests
rspec './spec/rubocop/target_finder_spec.rb[1:1:3:1]' # RuboCop::TargetFinder#find(..., :only_recognized_file_types) when files with a ruby extension are passed picks all the ruby files
rspec './spec/rubocop/target_finder_spec.rb[1:1:3:2:1]' # RuboCop::TargetFinder#find(..., :only_recognized_file_types) when files with a ruby extension are passed when local AllCops/Include lists two patterns picks two files
rspec './spec/rubocop/target_finder_spec.rb[1:1:3:2:2:1]' # RuboCop::TargetFinder#find(..., :only_recognized_file_types) when files with a ruby extension are passed when local AllCops/Include lists two patterns when a subdirectory AllCops/Include only lists one pattern picks two files
rspec './spec/rubocop/target_finder_spec.rb[1:1:4:1]' # RuboCop::TargetFinder#find(..., :only_recognized_file_types) when a file with a ruby filename is passed picks all the ruby files
rspec './spec/rubocop/target_finder_spec.rb[1:1:17:1]' # RuboCop::TargetFinder#find(..., :only_recognized_file_types) when input is passed on stdin includes the file
rspec './spec/rubocop/target_finder_spec.rb[1:2:13:1]' # RuboCop::TargetFinder#find(..., :all_file_types) when a non-ruby file is passed picks the file
rspec './spec/rubocop/target_finder_spec.rb[1:2:14:1]' # RuboCop::TargetFinder#find(..., :all_file_types) when files with a ruby extension are passed picks all the ruby files
rspec './spec/rubocop/target_finder_spec.rb[1:2:14:2:1]' # RuboCop::TargetFinder#find(..., :all_file_types) when files with a ruby extension are passed when local AllCops/Include lists two patterns picks all the ruby files
rspec './spec/rubocop/target_finder_spec.rb[1:2:14:2:2:1]' # RuboCop::TargetFinder#find(..., :all_file_types) when files with a ruby extension are passed when local AllCops/Include lists two patterns when a subdirectory AllCops/Include only lists one pattern picks all the ruby files
rspec './spec/rubocop/target_finder_spec.rb[1:2:16:1]' # RuboCop::TargetFinder#find(..., :all_file_types) when input is passed on stdin includes the file
rspec './spec/rubocop/target_finder_spec.rb[1:2:1:1]' # RuboCop::TargetFinder#find(..., :all_file_types) when a file with a ruby filename is passed picks all the ruby files
rspec './spec/ruby_lsp/rubocop/addon_spec.rb[1:2:1]' # RubyLSP::RuboCop::Addon textDocument/diagnostic has basic result information
rspec './spec/ruby_lsp/rubocop/addon_spec.rb[1:2:3]' # RubyLSP::RuboCop::Addon textDocument/diagnostic has second diagnostic information
rspec './spec/ruby_lsp/rubocop/addon_spec.rb[1:2:2]' # RubyLSP::RuboCop::Addon textDocument/diagnostic has first diagnostic information
rspec './spec/ruby_lsp/rubocop/addon_spec.rb[1:3:1]' # RubyLSP::RuboCop::Addon textDocument/formatting has basic result information
rspec './spec/ruby_lsp/rubocop/addon_spec.rb[1:3:2]' # RubyLSP::RuboCop::Addon textDocument/formatting has autocorrected code
```

---

Right, that makes sense. It probably wants to show a nice error when the user passes a file that doesn't exist instead of just silently ignoring it.

What do you think about `files.reject! { |file| FileTest.directory?(file) }`, that seems to pass all the tests.

---

I think it should be fine, thank you